pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java

```diff
@@ -130,6 +132,10 @@ public String toString() {
      * @return the decoded key/value pair
      */
     public static <K, V> KeyValue<K, V> decode(byte[] data, KeyValueDecoder<K, V> decoder) {
+        if (data == null || data.length == 0) {
```

Comment:
can you please explain how it can happen that we fall into these cases ?

I mean, probably data == null is possible in case of a message will NULL value, but I cannot think about a case in which the array may be of zero length.

probably in case of zero length array we should throw an IllegalArgumentException, because it is not possible that it is a valid value, as we are always writing the length of the key and the length of the value (with -1 as marker for nulls)

Reply:
It appends when key is SEPARATED and value is null, here is the KeyValueSchema.encode():

```
public byte[] encode(KeyValue<K, V> message) {
        if (keyValueEncodingType != null && keyValueEncodingType == KeyValueEncodingType.INLINE) {
            return KeyValue.encode(
                message.getKey(),
                keySchema,
                message.getValue(),
                valueSchema
            );
        } else {
            if (message.getValue() == null) {
                return null;
            }
            return valueSchema.encode(message.getValue());
        }
    }
```

Reply:
@vroyer I still do not see how we can have a zero byte array from your example.

if we have `keyValueEncodingType == KeyValueEncodingType.SEPARATED` then we are not using the KeyValue encoding,  but the value contains only the serialized version of the value (`valueSchema.encode(message.getValue())` ).

this "decode" code should be used only when you read an KeyValueEncodingType.INLINE message, that contains both the key and the value inside the payload of the message:
- keylen
- key bytes
- valuelen
- value bytes

when you serialize a null KeyValue you can a null value
when you serialize a null key you see -1 as "keylen" 
when you serialize a null value you see -1 as "valuelen"

there is no way to see a zero byte array

Reply:
Yes, https://github.com/apache/pulsar/pull/10089 is possibly the root cause. The KeyValue with a null value was decoded as INLINED, so this protection https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/KeyValueSchema.java#L146 was not working as expected.